extstore.c

```diff
@@ -710,6 +812,494 @@ static inline int _read_from_wbuf(store_page *p, obj_io *io) {
+    } else if (data->io->iov) {
+        size_t to_skip = bytes;
+
+        while (to_skip) {
+            size_t len = MIN(data->io->iov[data->iov_index].iov_len - data->iov_offset, to_skip);
+
+            data->iov_offset += len;
+            to_skip -= len;
+            if (data->iov_offset == data->io->iov[data->iov_index].iov_len && data->iov_index < data->io->iovcnt - 1) {
+                data->iov_index++;
+                data->iov_offset = 0;
+            }
+        }
+
+        data->iov_oneshot.iov_base = (char *)data->io->iov[data->iov_index].iov_base + data->iov_offset;
+        data->iov_oneshot.iov_len = data->io->iov[data->iov_index].iov_len - data->iov_offset;
+        data->iov = &data->iov_oneshot;
+        data->iovcnt = 1;
+    } else {
+        data->iov_offset += bytes;
+        data->iov_oneshot.iov_base = data->io->buf + data->iov_offset;
+        data->iov_oneshot.iov_len = data->end_offset - data->start_offset;
+    }
+
+    return data->start_offset == data->end_offset;
+}
+
+static void store_io_uring_submit_prep(struct store_io_data *data, struct io_uring_sqe *sqe) {
+    if (data->io->mode == OBJ_IO_READ) {
+        io_uring_prep_readv(sqe, data->fd, data->iov, data->iovcnt, data->start_offset);
+    } else {
+        io_uring_prep_writev(sqe, data->fd, data->iov, data->iovcnt, data->start_offset);
+    }
+
+    io_uring_sqe_set_data(sqe, data);
+}
+
+static bool store_io_uring_submit(store_io_thread *t, obj_io *io) {
+    struct store_io_uring_thread *priv = t->priv;
+    struct store_io_data *data = LIST_FIRST(&priv->io_free);
+    struct io_uring_sqe *sqe;
+    int ret;
+
+    if (!data) {
+        E_DEBUG("no free io data\n");
+        ret = -1;
+        extstore_io_done(t, io, ret);
+        return false;
+    }
+
+    LIST_REMOVE(data, link);
+
+    sqe = io_uring_get_sqe(&priv->ring);
+    if (!sqe) {
+        E_DEBUG("get sqe failed\n");
+        ret = -1;
+        goto error;
+    }
+
+    store_io_data_init(data, t->e, io);
+    store_io_uring_submit_prep(data, sqe);
+
+    LIST_INSERT_HEAD(&priv->io_inflight, data, link);
+
+    ret = io_uring_submit(&priv->ring);
+    if (ret < 0) {
+        E_DEBUG("io_uring_submit: %s\n", strerror(-ret));
+        LIST_REMOVE(data, link);
+        ret = -1;
+        goto error;
+    }
+
+    return true;
+
+error:
+    extstore_io_done(t, io, ret);
+    LIST_INSERT_HEAD(&priv->io_free, data, link);
+
+    return false;
+}
+
+static bool store_io_uring_complete(store_io_thread *t) {
+    struct store_io_uring_thread *priv = t->priv;
+    struct io_uring_cqe *cqe;
+    struct store_io_data *data;
+    int ret;
+    bool done;
+
+    if (LIST_EMPTY(&priv->io_inflight)) {
+        return false;
+    }
+
+    ret = io_uring_wait_cqe(&priv->ring, &cqe);
+    if (ret < 0) {
+        E_DEBUG("io_uring_wait_cqe: %s\n", strerror(-ret));
+        return false;
+    }
+
+    data = io_uring_cqe_get_data(cqe);
+
+    ret = cqe->res;
+    if (ret <= 0) {
```

Comment:
Presumably 'ret == 0' can only happen when reading the end of file (e.g. the extstore is accidentally truncated).
So I would like to avoid infinite loop by this condition.